runtime/ts/storage/firebase-storage.ts

```diff
@@ -1160,17 +1162,21 @@ class FirebaseCursor {
     const lastEntry = await getSnapshot(this.orderByIndex.limitToLast(1));
     lastEntry.forEach(entry => this.end = entry.val().index);
 
-    // Read one past the page size each time to establish the starting index for the next page.
-    this.baseQuery = this.orderByIndex.endAt(this.end).limitToFirst(this.pageSize + 1);
+    // Read one past the page size each time to establish the boundary index for the next page.
+    this.baseQuery = this.forward
+        ? this.orderByIndex.limitToFirst(this.pageSize + 1)
+        : this.orderByIndex.limitToLast(this.pageSize + 1);
 
     // Attach a listener for removed items and capture any that occur ahead of our streaming
     // frame. These will be returned after the cursor reaches the item at this.end.
     this.removedFn = this.orderByIndex.on('child_removed', snapshot => {
-      if (snapshot.val().index <= this.end &&
-          (this.nextStart === null || snapshot.val().index >= this.nextStart)) {
+      const index = snapshot.val().index;
+      if (index > this.end) return;
+      if (this.nextBoundary === null || (this.forward && index >= this.nextBoundary)
+                                     || (!this.forward && index <= this.nextBoundary)) {
         this.removed.push(snapshot.val().value);
       }
-    });
+    }); 
```

Comment:
nit: maybe configure your editor to clear trailing spaces on save?

Reply:
While I would prefer that, our code base has lots of trailing spaces in it.